### DISCUSSION TITLE
An alternative to 'chezmoi cd'

### DISCUSSION THREAD

**Discussion by Lockszmith-GH** [twpayne/chezmoi#3089] [Ideas]
Reading #3088, I wanted to ask if an alternative approach here wouldn't be better.

`chezmoi` is already providing shell specific code for tab-completion, so the concept isn't totally foreign.

Would it make sense to have chezmoi have an option to generate a native function in the shell that does the cd (or a pushd if available) without forking the shell.

I undertand why forking the shell is the 'easiest, most robust solution', but it's still spinning a new shell - and cding out of chezmoi dir, without exiting leaves those processes.

in bash it would be something like `. <( chezmoi util-funcs "cz" )` where it would created a `cz` funciton/alias to `chezmoi`, and `czcd` as a shell-native `chezmoi cd` replacement.

**Comment by halostatue** [comment]
It would be *possible*, but I don’t necessarily think that it would be a good idea. Chezmoi completions are provided by spf13/cobra and I don’t think that any additional work is done to maintain those for chezmoi (this is why `chezmoi cd<Tab>`doesn't complete subdirectories in `$CHEZMOIROOT`). This would now require that we maintain a library of shell function templates that could be used. That would include at a minimum (this is based on the shells that I have seen mentioned in the last year or so):

1. bash
2. zsh
3. fish
4. PowerShell
5. nushell

Yes, bash and zsh shell functions are mostly compatible, but they need not be as zsh is strictly more capable.

If this were accepted I would:
- recommend `shell-utils` as the command name;
- always require a shell parameter (either `chezmoi shell-utils --shell bash` or `chezmoi shell-utils bash`);
- consider whether a separate `*cd` function is required, because once you’re in a function, you can *always* override a subcommand (but that makes the functions themselves more complex);
- consider whether additional parameters should be accepted that would permit additional parameters to be passed to the invocation of `chezmoi` in the function, automating multirepo support.

I don’t think that I’d use this, and I think that support efforts would increase unless `chezmoi` had a way of detecting whether it was called from a shell alias (e.g., `chezmoi --__util_func "${@}"`) particularly for improvements to `chezmoi doctor`.

**Comment by twpayne** [comment]
The function for most shells is ([ref](https://www.chezmoi.io/user-guide/frequently-asked-questions/design/#why-does-chezmoi-cd-spawn-a-shell-instead-of-just-changing-directory)):

```bash
chezmoi-cd() {
    cd $(chezmoi source-path)
}
```

I'm not sure this is work generating.

Note that chezmoi already has a [`generate` command](https://www.chezmoi.io/reference/commands/generate/) that could be used to generate the function if we really wanted this.

**Reply by Lockszmith-GH** [reply]
Right, this is what I'm doing today.

I'm glad to see you're not opposed to it.

I'm thinking something along the lines of:

```bash
. <( chezmoi generate shell-utils bash "cz" )
```
Where the last parameter is an optional shorthand alias.

I would generate the following code:
```bash
cz() {
  chezmoi "${@}"
}

czcd() {
  cd "$(cz source-path ${1})"
}
```

While a PowerShell version (here without the alias) would generate:

```bash
. <( chezmoi generate shell-utils powershell )
```

```PowerShell
function chezmoi-cd {
param([string]$Path="")
    Set-Location $(& chezmoi source-path $Path)
}
```

**Reply by halostatue** [reply]
Note that `cd "$(chezmoi source-path)/$1"` is *not* the same as `chezmoi cd`. I can `chezmoi cd ~/.config` and it will go to `/home/halostatue/.local/share/chezmoi/home/private_dot_config`. However, `cd "$(chezmoi source-path "$1")"` will do the right thing, but detecting and reporting any error (`chezmoi: /home/halostatue/.configx: not managed`) may not be easy.

**Reply by Lockszmith-GH** [reply]
Fair enough, I didn't really expect the funciton to be a complete `chezmoi cd` replacement, if that is wanted, it can probably be done by modifying `chezmoi source-path` functionality to accept parameters and rely on it's output completlely.

**Reply by halostatue** [reply]
It already does (accept parameters), and on *bash* at least, the error output will show up even in an assignment:

```console
$ foo=$(chezmoi source-path .configx)
chezmoi: .configx: not managed
```

**Reply by Lockszmith-GH** [reply]
sweet, I've updated the example above. Thanks for looking into this
